### PR TITLE
Mapping state serialization

### PR DESCRIPTION
**Pull request by sebastianklose**
Adds a method that serialized the mapping state (PoseGraph with options) using the new serialization format defined in [RFC 0021](https://github.com/sebastianklose/rfcs/blob/serialization_format/text/0021-serialization-format.md)

Function not called from `MapBuilder` yet, will hook up in a separate PR once migration tool and deserialization helper is merged as well.

### CODE REVIEW

File: cartographer/io/internal/mapping_state_serialization.cc [hunk 1]
```diff
@@ -0,0 +1,171 @@
...
+ * limitations under the License.
+ */
+
+#include "cartographer/io/internal/mapping_state_serialization.h"
+#include "cartographer/mapping/proto/serialization.pb.h"
+#include "cartographer/transform/transform.h"
+
+namespace cartographer {
+namespace io {
+
```
Comment (pifon2a) [inline cartographer/io/internal/mapping_state_serialization.cc]: super-nit: remove the empty line between `namespace io {` and `namespace {`.

File: cartographer/io/internal/mapping_state_serialization.h [hunk 2]
```diff
@@ -0,0 +1,40 @@
...
+ * limitations under the License.
+ */
+#ifndef CARTOGRAPHER_IO_INTERNAL_MAPPING_STATE_SERIALIZATION_H_
+#define CARTOGRAPHER_IO_INTERNAL_MAPPING_STATE_SERIALIZATION_H_
+
+#include "cartographer/io/proto_stream_interface.h"
+#include "cartographer/mapping/pose_graph.h"
+#include "cartographer/mapping/proto/trajectory_builder_options.pb.h"
+
+// Helper function to serialize a PoseGraph to a protobuf stream.
```
Comment (pifon2a) [inline cartographer/io/internal/mapping_state_serialization.h]: nit: it looks like this comment belongs to `ToPbStream` func.  I do not think there is a need for a file-level comment. Up to you.

File: cartographer/io/internal/mapping_state_serialization.h [hunk 3]
```diff
@@ -0,0 +1,40 @@
...
+namespace cartographer {
+namespace io {
+
+// The current serialization format version.
+static constexpr int kMappingStateSerializationFormatVersion = 1;
```
Comment (pifon2a) [inline cartographer/io/internal/mapping_state_serialization.h]: will the version number be used outside of `mapping_state_serialization.cc`?

File: cartographer/io/internal/mapping_state_serialization.cc [hunk 4]
```diff
@@ -0,0 +1,171 @@
...
+  // Next serialize all submaps.
+  for (const auto& submap_id_data : pose_graph.GetAllSubmapData()) {
+    auto* const submap_proto = proto.mutable_submap();
+    submap_proto->mutable_submap_id()->set_trajectory_id(
+        submap_id_data.id.trajectory_id);
+    submap_proto->mutable_submap_id()->set_submap_index(
+        submap_id_data.id.submap_index);
+    submap_id_data.data.submap->ToProto(submap_proto,
+                                        /*include_probability_grid_data=*/true);
+    writer->WriteProto(proto);
```
Comment (pifon2a) [inline cartographer/io/internal/mapping_state_serialization.cc]: I would wrap these blocks into functions, that way there won't be a need for comments and also for {} blocks.

File: cartographer/io/internal/mapping_state_serialization.cc (hunk 1 above)
Comment (sebastianklose) [reply cartographer/io/internal/mapping_state_serialization.cc]: done.

File: cartographer/io/internal/mapping_state_serialization.cc (hunk 4 above)
Comment (sebastianklose) [reply cartographer/io/internal/mapping_state_serialization.cc]: done

File: cartographer/io/internal/mapping_state_serialization.h (hunk 2 above)
Comment (sebastianklose) [reply cartographer/io/internal/mapping_state_serialization.h]: rm'ed

File: cartographer/io/internal/mapping_state_serialization.h (hunk 3 above)
Comment (sebastianklose) [reply cartographer/io/internal/mapping_state_serialization.h]: In a different branch I am using it to verify the version in the deserialization. We can also do it differently.

Comment (pifon2a) [reply cartographer/io/internal/mapping_state_serialization.h]: LGTM/WLDDTM (wubba lubba dub dub to me).

**Pull request merged**